### DISCUSSION TITLE
Generating meshes from CAD assemblies

### DISCUSSION THREAD

**Discussion by s-dunnim** [idaholab/moose#22032] [Q&A Meshing] [answered]
Hi,

I'm working on a MOOSE model which will require external CAD to be meshed. The CAD is provided as an assembly in a STEP file; and I would like to make it into a single, continuous, high quality, tetrahedral mesh. Does anyone have any experience/recommendations for a pipeline/process to achieve this?

So far I have tried:
- Using blender to "boolean" all the parts together, export to STL, and mesh in tetgen. I struggled to avoid intersecting segments/facets, and the conversion to STL discretises the geometry before meshing, often with "ugly" triangles.
- Using FreeCAD to "fuse" parts together, export to STL, and mesh in tetgen. This comes with the same problems as above, as well as the fact FreeCAD struggles to fuse parts which aren't exactly aligned (in Blender I was doing a little manual extrusion to fix this).

I haven't tried:
- Hand coding a script to make a poly file for tetgen, as the geometry is too complex.
- Using Cubit as discussed in this [post](https://github.com/idaholab/moose/discussions/17884), as it isn't free for commercial use.
- Using Gmsh. This is likely our next port of call, especially if it is recommended here.

Thanks.

**Comment by GiudGiud** [comment, marked as answer]
Hello

How complicated is the CAD? We have some native support in MOOSE for simple shapes, and we can stitch, rotate etc
@grmnptr would SALOME be an option here?

Guillaume

**Reply by s-dunnim** [reply]
Hi, thanks for the response.
It's fairly complex, with fillets, chamfers, round edges, non-planar surfaces etc. I looked into maybe using StitchedMesh, but I can't guarantee the faces I'm stitching will have corresponding nodes, or even be exactly aligned. It also doesn't solve the issue of meshing from CAD without an intermediary STL conversion generating unwanted triangles. I'll look into SALOME.

**Reply by grmnptr** [reply]
Both SALOME and gmsh can do this; both have multiple algorithms and python APIs for a fine control over the meshing process. I personally use gmsh for simpler meshes and SALOME for more complex ones. However, I have to admit I haven't tried any of these for meshing such complex geometries. I think SALOME's GUI is nicer, not sure if that matters.

**Reply by s-dunnim** [reply]
Thank you for the advice, I'll have a go with those tools and see what I can do! Will come back and mark as answer if it works out for me.
